File: api/sources/yahoo.py
"""Yahoo Finance chart endpoint — for global indices/futures (DJI, WTI, YM)."""
from __future__ import annotations
from .http import client

CHART_URL = "https://query1.finance.yahoo.com/v8/finance/chart/{symbol}"


async def get_chart(symbol: str, range_: str = "1d", interval: str = "5m") -> dict:
    """Return normalized chart + quote summary for a Yahoo symbol.

    Examples: '^DJI', 'CL=F' (WTI futures), 'YM=F' (DJ futures).
    """
    r = await client().get(
        CHART_URL.format(symbol=symbol),
        params={"range": range_, "interval": interval},
    )
    r.raise_for_status()
    payload = r.json()
    results = (payload.get("chart") or {}).get("result") or []
    if not results:
        return _empty(symbol)
    res = results[0]
    meta = res.get("meta") or {}
    ts: list[int] = res.get("timestamp") or []
    quote = ((res.get("indicators") or {}).get("quote") or [{}])[0]

    closes = quote.get("close") or []
    opens = quote.get("open") or []
    highs = quote.get("high") or []
    lows = quote.get("low") or []

    price = meta.get("regularMarketPrice")
    prev_close = meta.get("chartPreviousClose") or meta.get("previousClose")
    change = round(price - prev_close, 4) if price is not None and prev_close else None
    change_pct = (
        round(change / prev_close * 100, 4)
        if change is not None and prev_close
        else None
    )

    # filter out null candles
    pairs = [(t, c) for t, c in zip(ts, closes) if c is not None]
    t_clean = [p[0] for p in pairs]
    c_clean = [round(p[1], 4) for p in pairs]

    return {
        "symbol": symbol,
        "name": meta.get("shortName") or meta.get("longName") or symbol,
        "price": round(price, 4) if price is not None else None,
        "prev": round(prev_close, 4) if prev_close else None,
        "change": change,
        "changePct": change_pct,
        "currency": meta.get("currency"),
        "t": t_clean,
        "c": c_clean,
        "o": [round(v, 4) if v is not None else None for v in opens],
        "h": [round(v, 4) if v is not None else None for v in highs],
        "l": [round(v, 4) if v is not None else None for v in lows],
    }


def _empty(symbol: str) -> dict:
    return {
        "symbol": symbol,
        "name": symbol,
        "price": None,
        "prev": None,
        "change": None,
        "changePct": None,
        "currency": None,
        "t": [],
        "c": [],
        "o": [],
        "h": [],
        "l": [],
    }

File: web/src/lib/server/yahoo.ts
const UA =
  "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/126.0 Safari/537.36";

export type YahooChart = {
  symbol: string;
  name: string;
  price: number | null;
  prev: number | null;
  change: number | null;
  changePct: number | null;
  t: number[];
  c: number[];
};

export async function getChart(
  symbol: string,
  range_ = "1d",
  interval = "5m"
): Promise<YahooChart> {
  const url = `https://query1.finance.yahoo.com/v8/finance/chart/${encodeURIComponent(symbol)}?range=${range_}&interval=${interval}`;
  try {
    const r = await fetch(url, {
      headers: { "User-Agent": UA, Accept: "application/json" },
      next: { revalidate: 30 },
    });
    if (!r.ok) return empty(symbol);
    const payload = await r.json();
    const results: Record<string, unknown>[] = payload?.chart?.result ?? [];
    if (!results.length) return empty(symbol);

    const res = results[0];
    const meta = (res.meta ?? {}) as Record<string, unknown>;
    const ts: number[] = (res.timestamp as number[]) ?? [];
    const quote = ((res.indicators as Record<string, unknown[]>)?.quote ?? [{}])[0] as Record<string, (number | null)[]>;

    const closes = quote.close ?? [];
    const price = meta.regularMarketPrice as number | null;
    const prevClose = (meta.chartPreviousClose ?? meta.previousClose) as number | null;
    const change = price != null && prevClose ? Math.round((price - prevClose) * 10000) / 10000 : null;
    const changePct = change != null && prevClose ? Math.round((change / prevClose) * 100 * 10000) / 10000 : null;

    const pairs = ts.map((t, i) => [t, closes[i]] as [number, number | null]).filter(([, c]) => c != null);

    return {
      symbol,
      name: (meta.shortName ?? meta.longName ?? symbol) as string,
      price: price != null ? Math.round(price * 10000) / 10000 : null,
      prev: prevClose != null ? Math.round(prevClose * 10000) / 10000 : null,
      change,
      changePct,
      t: pairs.map(([t]) => t),
      c: pairs.map(([, c]) => Math.round(c! * 10000) / 10000),
    };
  } catch {
    return empty(symbol);
  }
}

function empty(symbol: string): YahooChart {
  return { symbol, name: symbol, price: null, prev: null, change: null, changePct: null, t: [], c: [] };
}
